File: inventory-service/src/main/java/com/smartcart/inventory/service/InventoryService.java
package com.smartcart.inventory.service;

import com.smartcart.common.dto.PageResponse;
import com.smartcart.common.event.OrderCancelledEvent;
import com.smartcart.common.event.OrderCreatedEvent;
import com.smartcart.common.event.PaymentFailedEvent;
import com.smartcart.common.event.PaymentSuccessEvent;
import com.smartcart.inventory.dto.*;

import java.util.List;

public interface InventoryService {

    InventoryResponse addInventory(InventoryRequest request);

    InventoryResponse restock(String sku, RestockRequest request);

    InventoryResponse getInventoryBySku(String sku);

    List<StockCheckResponse> batchCheckStock(List<String> skus);

    PageResponse<InventoryResponse> getAllInventory(int page, int size);

    PageResponse<InventoryResponse> getLowStockInventory(int page, int size);

    void processOrderCreated(OrderCreatedEvent event);

    void processPaymentSuccess(PaymentSuccessEvent event);

    void processPaymentFailed(PaymentFailedEvent event);

    void processOrderCancelled(OrderCancelledEvent event);
}

File: smartcart-frontend/src/app/core/services/inventory.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, map } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApiResponse, PageResponse } from '../models/api.models';
import { InventoryRequest, InventoryResponse, RestockRequest } from '../models/inventory.models';

@Injectable({
  providedIn: 'root'
})
export class InventoryService {
  private http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/api/v1/inventory`;

  getInventoryBySku(sku: string): Observable<InventoryResponse> {
    return this.http.get<ApiResponse<InventoryResponse>>(`${this.baseUrl}/${sku}`).pipe(
      map(res => res.data)
    );
  }

  getAllInventory(page: number = 0, size: number = 10): Observable<PageResponse<InventoryResponse>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    return this.http.get<ApiResponse<PageResponse<InventoryResponse>>>(this.baseUrl, { params }).pipe(
      map(res => res.data)
    );
  }

  getLowStockInventory(page: number = 0, size: number = 10): Observable<PageResponse<InventoryResponse>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    return this.http.get<ApiResponse<PageResponse<InventoryResponse>>>(`${this.baseUrl}/low-stock`, { params }).pipe(
      map(res => res.data)
    );
  }

  addInventory(request: InventoryRequest): Observable<InventoryResponse> {
    return this.http.post<ApiResponse<InventoryResponse>>(this.baseUrl, request).pipe(
      map(res => res.data)
    );
  }

  restock(sku: string, quantity: number): Observable<InventoryResponse> {
    const payload: RestockRequest = { quantity };
    return this.http.put<ApiResponse<InventoryResponse>>(`${this.baseUrl}/${sku}/restock`, payload).pipe(
      map(res => res.data)
    );
  }
}
